# 01. Working with Data Using Pandas
## 📚 Learning Objectives

By completing this notebook, you will:
- Work with Pandas DataFrames effectively
- Manipulate and transform data
- Filter and select data
- Aggregate and group data
- Apply common Pandas operations for ML preparation

## 🔗 Where this fits

**Builds on:** Course 05 — Unit 1, lesson 02 (pandas basics) and Course 04 (AIAT 114) — Unit 1: this unit deliberately revisits Course 04's algorithms, and the difference is the data-science workflow wrapped around them.

**Used later in:** Course 05 — Unit 4, lesson 02, which turns these manipulations into a model-ready table.

---

This notebook covers practical activities from **Course 05, Unit 4**:
- Working with data using Python libraries like Pandas

---

## The Story: Preparing Your Workspace

Imagine you're building furniture. **Before** assembling, you organize your workspace - arrange tools, prepare materials, set up workbench. **After** organizing, you can work efficiently!

Same with ML: **Before** building models, we manipulate data - filter, transform, aggregate. **After** manipulating, data is ready for models!

---

## Why Pandas Data Manipulation Matters

Pandas data manipulation is essential because:
- **Foundation**: All ML work starts with data manipulation
- **Preparation**: Manipulate data into ML-ready format
- **Efficiency**: Pandas is fast and powerful
- **Standard**: Industry standard for data manipulation

**Common Student Questions:**
- **Q: What's the difference between loc and iloc?**
  - Answer: loc uses labels, iloc uses integer positions
  - Example: df.loc[0] = row with index 0, df.iloc[0] = first row
  - Tip: Use loc for labels, iloc for positions
  
- **Q: When do I use groupby?**
  - Answer: When you need to aggregate by categories
  - Example: Group by 'category', then calculate mean
  - Benefit: Powerful for summarizing data

---

## Introduction

**Pandas** is the standard Python library for data manipulation and analysis. This notebook demonstrates essential Pandas operations for preparing data for machine learning tasks.


## 🎯 The case: Berkeley, 1973 — a bias that vanished when someone used groupby

The University of California, Berkeley graduate admissions for autumn 1973 looked
indefensible. Of **8,442 male applicants, 44% were admitted**; of **4,321 women, 35%**.
Across 12,763 applicants the gap was far too large to be chance, and the university faced
the obvious accusation.

Then Bickel, Hammel and O'Connell broke the same numbers down **by department** and
published the result in *Science* in 1975 (*Sex Bias in Graduate Admissions: Data From
Berkeley*, 187:398–404). Department by department, there was a small bias **in favour of
women**. Women had applied disproportionately to the departments with the lowest admission
rates for everybody; men had applied to the ones with the highest. The aggregate reversed
the direction of every one of its parts.

This is **Simpson's paradox**, and the only tool that exposes it is the one this notebook
is about: `groupby`. The single number was not wrong. It was an average over groups that
should never have been averaged.

**What goes wrong without this.** You are about to compute that first-class women survived
the Titanic at **96.8%** and third-class men at **13.5%**, from two filters. Report only
the overall 38.4% survival rate and both facts disappear. The aggregate is always
available, always defensible, and frequently the least informative number in the table.


## 📥 Inputs & 📤 Outputs

**Inputs:** What we use in this notebook

- **Real data:** `titanic.csv` — the 891-passenger manifest of the RMS Titanic
  (name, sex, age, travel class, ticket fare, cabin, port of embarkation and
  whether the passenger survived). It carries real gaps: 177 missing ages,
  687 missing cabins, 2 missing embarkation ports.
- pandas

**Outputs:** What you'll see when you run the cells

- Selections, filters, groupings and string operations run on real passenger records
- Every printed number is computed from the manifest, not typed in

---


In [1]:
# Imports
import pandas as pd
import numpy as np

In [2]:
# WHAT: Load the real Titanic passenger manifest.
# WHY: Every pandas skill below - selecting, filtering, grouping, string work -
#      is worth learning because real tables are messy; this one has 177 missing ages.

print("PART 1: Load the data")
print("-" * 70)

DATA_DIR = '../../../Course 04/datasets/raw/'
df = pd.read_csv(DATA_DIR + 'titanic.csv')

print("Real Titanic manifest (first 6 rows):")
print(df.head(6).to_string())
print(f"\nDataFrame shape: {df.shape}")
print("DataFrame info:")
print(df.info())
print(f"\nMissing values per column (real gaps, nothing removed):")
print(df.isnull().sum()[df.isnull().sum() > 0].to_string())

# ============================================================================

PART 1: Load the data
----------------------------------------------------------------------
Real Titanic manifest (first 6 rows):
   PassengerId  Survived  Pclass                                                 Name     Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked
0            1         0       3                              Braund, Mr. Owen Harris    male  22.0      1      0         A/5 21171   7.2500   NaN        S
1            2         1       1  Cumings, Mrs. John Bradley (Florence Briggs Thayer)  female  38.0      1      0          PC 17599  71.2833   C85        C
2            3         1       3                               Heikkinen, Miss. Laina  female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S
3            4         1       1         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1      0            113803  53.1000  C123        S
4            5         0       3                             Allen, Mr. William Henry    

In [3]:
# WHAT: Select rows and columns two ways - loc (by label) and iloc (by position).
# WHY: Mixing the two up is the most common pandas bug; doing both side by side fixes it early.

# ============================================================================
print("\n" + "=" * 70)
print("PART 2: Data Selection")
print("=" * 70)

# Using loc (label-based)
print("\n✅ Example 1: Using loc (Label-based Selection)")
print("-" * 70)
print("Select rows by label (index 0-2):")
print(df.loc[0:2, ['Name', 'Sex', 'Age', 'Pclass']].to_string())
print("\nSelect specific columns:")
print(df.loc[:, ['Name', 'Age', 'Fare']].head().to_string())
print("\nSelect rows and columns:")
print(df.loc[0:2, ['Name', 'Fare']].to_string())

# Using iloc (integer position-based)
print("\n✅ Example 2: Using iloc (Position-based Selection)")
print("-" * 70)
print("Select first 3 rows:")
print(df.iloc[0:3, 0:6].to_string())
print("\nSelect specific columns by position (Survived, Pclass, Name):")
print(df.iloc[:, [1, 2, 3]].head().to_string())

# ============================================================================


PART 2: Data Selection

✅ Example 1: Using loc (Label-based Selection)
----------------------------------------------------------------------
Select rows by label (index 0-2):
                                                  Name     Sex   Age  Pclass
0                              Braund, Mr. Owen Harris    male  22.0       3
1  Cumings, Mrs. John Bradley (Florence Briggs Thayer)  female  38.0       1
2                               Heikkinen, Miss. Laina  female  26.0       3

Select specific columns:
                                                  Name   Age     Fare
0                              Braund, Mr. Owen Harris  22.0   7.2500
1  Cumings, Mrs. John Bradley (Florence Briggs Thayer)  38.0  71.2833
2                               Heikkinen, Miss. Laina  26.0   7.9250
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  35.0  53.1000
4                             Allen, Mr. William Henry  35.0   8.0500

Select rows and columns:
                                           

In [4]:
# WHAT: Filter rows with boolean masks - single and combined conditions.
# WHY: Filtering is how you ask a table a question; every analysis below starts with one.

# ============================================================================
print("\n" + "=" * 70)
print("PART 3: Data Filtering")
print("=" * 70)

# Boolean indexing
print("\n✅ Example 3: Boolean Indexing (Filtering)")
print("-" * 70)
first_class = df[df['Pclass'] == 1]
print(f"First-class passengers: {len(first_class)} of {len(df)}")
print(first_class[['Name', 'Sex', 'Age', 'Fare']].head().to_string())

expensive = df[df['Fare'] > 100]
print(f"\nPassengers who paid more than £100: {len(expensive)}")
print(f"  Their survival rate: {expensive['Survived'].mean():.1%}")
print(f"  Everyone else's survival rate: {df[df['Fare'] <= 100]['Survived'].mean():.1%}")

# Multiple conditions
print("\n✅ Example 4: Multiple Conditions")
print("-" * 70)
women_first = df[(df['Sex'] == 'female') & (df['Pclass'] == 1)]
print(f"Women in first class: {len(women_first)}, "
      f"survival rate {women_first['Survived'].mean():.1%}")
men_third = df[(df['Sex'] == 'male') & (df['Pclass'] == 3)]
print(f"Men in third class:  {len(men_third)}, "
      f"survival rate {men_third['Survived'].mean():.1%}")
print("\n💡 Two filters just uncovered the single starkest pattern in the manifest.")

# ============================================================================


PART 3: Data Filtering

✅ Example 3: Boolean Indexing (Filtering)
----------------------------------------------------------------------
First-class passengers: 216 of 891
                                                   Name     Sex   Age     Fare
1   Cumings, Mrs. John Bradley (Florence Briggs Thayer)  female  38.0  71.2833
3          Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0  53.1000
6                               McCarthy, Mr. Timothy J    male  54.0  51.8625
11                             Bonnell, Miss. Elizabeth  female  58.0  26.5500
23                         Sloper, Mr. William Thompson    male  28.0  35.5000

Passengers who paid more than £100: 53
  Their survival rate: 73.6%
  Everyone else's survival rate: 36.2%

✅ Example 4: Multiple Conditions
----------------------------------------------------------------------
Women in first class: 94, survival rate 96.8%
Men in third class:  347, survival rate 13.5%

💡 Two filters just uncovered the single starkes

In [5]:
# WHAT: Add computed columns, then group and aggregate.
# WHY: groupby answers 'how does this differ between groups?' - the workhorse of analysis.

# ============================================================================
print("\n" + "=" * 70)
print("PART 4: Data Manipulation")
print("=" * 70)

# Adding columns
print("\n✅ Example 5: Adding Columns")
print("-" * 70)
df['family_size'] = df['SibSp'] + df['Parch'] + 1   # siblings/spouses + parents/children + self
df['is_alone'] = (df['family_size'] == 1)
print(df[['Name', 'SibSp', 'Parch', 'family_size', 'is_alone']].head().to_string())
print(f"\nTravelling alone: {df['is_alone'].sum()} of {len(df)} passengers "
      f"({df['is_alone'].mean():.1%})")

# Grouping and aggregation
print("\n✅ Example 6: Grouping and Aggregation")
print("-" * 70)
class_stats = df.groupby('Pclass').agg({
    'Fare': ['mean', 'max', 'count'],
    'Age': 'mean',
    'Survived': 'mean'
}).round(2)
print("Statistics by travel class:")
print(class_stats.to_string())
print("\n💡 'Age' counts differ from 'Fare' counts because 177 ages are missing -")
print("   pandas skips NaN in mean() silently. Always check what you averaged over.")

# ============================================================================


PART 4: Data Manipulation

✅ Example 5: Adding Columns
----------------------------------------------------------------------
                                                  Name  SibSp  Parch  family_size  is_alone
0                              Braund, Mr. Owen Harris      1      0            2     False
1  Cumings, Mrs. John Bradley (Florence Briggs Thayer)      1      0            2     False
2                               Heikkinen, Miss. Laina      0      0            1      True
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)      1      0            2     False
4                             Allen, Mr. William Henry      0      0            1      True

Travelling alone: 537 of 891 passengers (60.3%)

✅ Example 6: Grouping and Aggregation
----------------------------------------------------------------------
Statistics by travel class:
         Fare                  Age Survived
         mean     max count   mean     mean
Pclass                                     
1 

In [6]:
# WHAT: Sort rows, then pull structure out of the free-text Name column with .str.
# WHY: String columns hide features - here the social title (Mr, Mrs, Miss, Master) is buried in a name.

# PART 5: DATA TRANSFORMATION
# ============================================================================
print("\n" + "=" * 70)
print("PART 5: Data Transformation")
print("=" * 70)

# Sorting
print("\n✅ Example 7: Sorting Data")
print("-" * 70)
sorted_by_fare = df.sort_values('Fare', ascending=False)
print("The five most expensive tickets on the manifest:")
print(sorted_by_fare[['Name', 'Pclass', 'Fare', 'Survived']].head().to_string())

# String operations
print("\n✅ Example 8: String Operations")
print("-" * 70)
# Names look like 'Braund, Mr. Owen Harris' - the title sits between ', ' and '.'
df['title'] = df['Name'].str.split(', ').str[1].str.split('.').str[0].str.strip()
df['name_length'] = df['Name'].str.len()
print("Extracted titles (top 6):")
print(df['title'].value_counts().head(6).to_string())
print("\nSurvival rate by extracted title (groups of 5+ passengers):")
title_counts = df['title'].value_counts()
big_titles = title_counts[title_counts >= 5].index
print(df[df['title'].isin(big_titles)].groupby('title')['Survived']
        .agg(['mean', 'size']).round(3).to_string())

# ============================================================================
# ============================================================================
print("\n" + "=" * 70)
print("Summary")
print("=" * 70)
print(f"""
✅ What you learned, on the real {len(df)}-passenger Titanic manifest:
   1. Data Selection: loc (labels), iloc (positions)
   2. Data Filtering: Boolean indexing, multiple conditions
      -> women in 1st class survived at {women_first['Survived'].mean():.0%},
         men in 3rd class at {men_third['Survived'].mean():.0%}
   3. Data Manipulation: computed columns (family_size), groupby, aggregate
   4. Data Transformation: sort, and .str to mine a title out of every name

🎯 Key Takeaways:
   - loc: Label-based selection (use index/column names)
   - iloc: Position-based selection (use integer positions)
   - Boolean indexing: Filter rows based on conditions
   - Groupby: Aggregate data by categories
   - Real data has gaps: {df['Age'].isna().sum()} ages are missing, and mean() ignores them silently

📚 Next Steps:
   - Example 2: Data Preparation for ML (we clean this same manifest for a model)
   - Example 3: Implementing ML Models
""")
print("✅ Pandas concepts understood!")


PART 5: Data Transformation

✅ Example 7: Sorting Data
----------------------------------------------------------------------
The five most expensive tickets on the manifest:
                                   Name  Pclass      Fare  Survived
258                    Ward, Miss. Anna       1  512.3292         1
737              Lesurer, Mr. Gustave J       1  512.3292         1
679  Cardeza, Mr. Thomas Drake Martinez       1  512.3292         1
88           Fortune, Miss. Mabel Helen       1  263.0000         1
27       Fortune, Mr. Charles Alexander       1  263.0000         0

✅ Example 8: String Operations
----------------------------------------------------------------------
Extracted titles (top 6):
title
Mr        517
Miss      182
Mrs       125
Master     40
Dr          7
Rev         6

Survival rate by extracted title (groups of 5+ passengers):
         mean  size
title              
Dr      0.429     7
Master  0.575    40
Miss    0.698   182
Mr      0.157   517
Mrs     0.792   

## ⚠️ Where this breaks

- **Every aggregate can reverse when you split it.** Simpson's paradox is not exotic; it
  is what happens whenever group sizes differ and the groups behave differently. There is
  no statistical test that tells you which level of aggregation is "correct" — that is a
  question about the causal structure, and it has to be answered outside the data.
- **`groupby` will happily average things that must not be averaged.** The mean of a
  ratio is not the ratio of means; the mean of `Survived` is a rate only because the column
  is 0/1; averaging a percentage across groups of unequal size is almost always wrong.
- **Chained assignment fails silently.** `df[df.Age > 30]['Fare'] = 0` may modify a copy
  and leave the original untouched — with a warning most people have learned to ignore.
  Use `.loc[mask, 'Fare'] = 0`.
- **`.str` extraction is regex on human text.** The titles came out clean here because a
  1912 manifest is formatted consistently. Real free-text fields have typos, mixed
  languages, missing separators and encodings. Always print the value counts of what you
  extracted before using it — including the tail.
- **The assumption that must hold: one row is one unit of the thing you are counting.**
  Unit 2 lesson 02 showed 210 passengers sharing ticket numbers. If you group by ticket
  you are counting *bookings*, not *people*, and every rate changes.
- **When pandas is the wrong tool:** if the data already lives in a database, do the
  grouping in SQL on the server. If the table is bigger than memory, Unit 5's Dask and
  Spark lessons exist for that. And if the answer needs a confidence interval rather than
  a point estimate, `groupby().mean()` is the start of the analysis, not the end.


## 💬 Discuss

`.str` extraction just pulled a social title out of free text and produced survival rates:
**Mrs 79.2% (n=125), Miss 69.8% (n=182), Master 57.5% (n=40), Dr 42.9% (n=7), Mr 15.7%
(n=517), Rev 0.0% (n=6)**.

1. "Rev: 0% survived" is a real number from six real clergymen. Would you put it in a
   report? State the rule you would apply for minimum group size — and then justify why
   "Dr, n=7" gets the same treatment or does not.
2. `title` is derived from `Name` and mostly encodes sex, age and marital status, which are
   already columns. So is it a new feature or a repackaged one? Argue whether adding it to
   a survival model is feature engineering or self-deception.
3. Berkeley's overall gap reversed when split by department. This manifest's survival rate
   is 38.4% overall. Name the split you would run *first* to check whether 38.4% is hiding
   a reversal — and say what result would make you stop trusting the headline number.


## 📚 References

1. McKinney, W. (2010). *Data Structures for Statistical Computing in Python*. Proceedings of the 9th Python in Science Conference (SciPy). <https://doi.org/10.25080/Majora-92bf1922-00a>
2. Wickham, H. (2014). *Tidy Data*. Journal of Statistical Software, 59(10), 1-23. <https://doi.org/10.18637/jss.v059.i10>
3. McKinney, W. (2017). *Python for Data Analysis*, 2nd ed. O'Reilly Media.